tdm/docs/architecture.md

```diff
@@ -0,0 +1,262 @@
+                                                       TEE-IO ABI
+                                                            │
+                                                            │
+                                                            ▽
+┏━━━━━━━━━━━━━━━━━━━━━━━┓                    ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+┃                       ┃                    ┃                            ┃
+┃  Host VMM/Hypervisor  ┃                    ┃  Trusted Security Manager  ┃
+┃    (KVM+QEMU/CLH)     ┃◁──────TEE ABI─────▷┃   (TDX Module, ASP, etc)   ┃
+┃                       ┃                    ┃                            ┃
+┗━━━━━━━━━━━━━━━━━━━━━━━┛                    ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+            △
+            ┃
+        PCIe Bus
+            ┃
+            ┃
+            ▽
+┏━━━━━━━━━━━━━━━━━━━━━━━┓
+┃ TEE-IO Compliant PCIe ┃
+┃        Device         ┃
+┗━━━━━━━━━━━━━━━━━━━━━━━┛
+
+```
+
+In order for the TDM to achieve the above described goals, it must be able to:
+
+1. Detect VMM-assigned PCIe devices. This relies on the Linux kernel `udev`
+   subsystem. TDM also assumes the TSM must not enable any DMA or MMIO mappings
+   before it accepted the device.
+2. Authenticate the assigned device. Both authentication and attestation of the
+   TDI relies on the ability for the TDM to retrieve a certificate chain and an
+   attestation evidence. This is provided by the Linux kernel confidential
+   computing ABI, which is generally exposed as a vendor-specific, ioctl-based
+   character device.
+3. Verify the assigned device. Before accepting a TDI into the confidential
+   guest TCB, the TDM must use the Linux kernel CoCo ABI to verify the following
+   security attributes from the TSM:
+   * Device link state
+   * TDI state
+   * TDI MMIO mappings
+4. Attest of the assigned device trustworthiness. As part of the overall device
+   verification process, the TDM must run a local or remote attestation of the
+   device attestation evidence:
+   1. For running local device attestation, the TDM needs to practically become a
+      Relying Party.
+   2. In order to support remote device attestation, the TDM would rely on
+      existing guest components like the attestation agent.
+5. Notify the TSM of it TDI acceptance decision. Here as well, this relies on
+   the Linux kernel CoCo ABI.
+
+This leads to a natural split of the TDM software architecture into 4 main
+modules:
+
+1. The **TDM Core** runs the main TDM flows and managing confidential guest
+   assignment with the support of the detection, ABI plugins and attestation
+   modules.
+2. The **Device Detection** module traps device discovery, assignment and
+   removal from the confidential guest PCIe emulated bus.
+3. The **Linux CoCo ABI plugins** abstracts the vendor specific CoCo Linux
+   kernel ABIs in order expose a vendor-agnostic internal interface for the TDM
+   core to consume.
+4. The **Device Attestation** module is responsible for verifying a device
```

Comment:
Note to myself (Thanks @jyao1): The attestation module may have to support vendor specific attestation results, so it may need to be modularized as well.

Reply:
We may leverage here: https://github.com/confidential-containers/confidential-containers/issues/135 
- Interoperability outside the CoCo flows - for example to reuse the AR computed by CoCo in subsequent TLS exchanges involving the CoCo attester,
- Ease the task of defining and computing authorization policies by relying parties due to the reliance on AR4SI's "trustworthiness vector" to present a normalized view of the evaluation results. An example showing how to use [OPA](https://www.openpolicyagent.org/) to evaluate a EAR in just a few lines of Rego can be found [here](https://play.openpolicyagent.org/p/fBydqNm0G0) (click on the Evaluate button.)
-  Code reuse via existing open-source libraries,
- Normalised and attester-agnostic result format, which allows to accommodate all currently supported formats as well as any other kinds of attester / CC technologies in the future.

Reply:
Oh, I'm all for it for sure. There may be cases where we would not talk to KBS but to vendor specific attestation services that return non EAR formatted attestation results.

Reply:
I treat [EAR](https://datatracker.ietf.org/doc/draft-fv-rats-ear/) to be **one** possible option.
But I am not sure if that is the **only** option.

Do we want to allow different format?